ServerCore/Pages/Submissions/Index.cshtml.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using ServerCore.DataModel;
+using ServerCore.ModelBases;
+
+namespace ServerCore.Pages.Submissions
+{
+    public class IndexModel : EventSpecificPageModel
+    {
+        private readonly ServerCore.Models.PuzzleServerContext _context;
+
+        public IndexModel(ServerCore.Models.PuzzleServerContext context)
+        {
+            _context = context;
+        }
+
+        [BindProperty]
+        public Submission Submission { get; set; }
+
+        public IList<Submission> Submissions { get; set; }
+
+        public int PuzzleId { get; set; }
+
+        public int TeamId { get; set; }
+
+        public string AnswerToken { get; set; }
+
+        public async Task<IActionResult> OnPostAsync(int puzzleId, int teamId)
+        {
+            if (!ModelState.IsValid)
+            {
+                return Page();
+            }
+
+            // Create submission and add it to list
+            Submission.TimeSubmitted = DateTime.Now;
+            Submission.Puzzle = await _context.Puzzles.SingleOrDefaultAsync(p => p.ID == puzzleId);
+            Submission.Team = await _context.Teams.Where((t) => t.ID == teamId).FirstOrDefaultAsync();
+
+            List<Response> responses = await _context.Responses.Where(r => r.Puzzle.ID == puzzleId && Submission.SubmissionText == r.SubmittedText).ToListAsync();
+            Submission.Response = responses.FirstOrDefault();
+
+            // Update puzzle state if submission was correct
+            if (Submission.Response != null && Submission.Response.IsSolution)
+            {
+                PuzzleStatePerTeam puzzleState = await _context.PuzzleStatePerTeam.Where(state => state.TeamID == teamId && state.PuzzleID == puzzleId).FirstOrDefaultAsync();
```

Comment:
Oops, missed this earlier - please use `PuzzleStateHelper.GetFullReadWriteQueryAsync` for this call, to match other uses. It's not strictly required in this case, but I'm trying to keep all the state code as similar as I can.